README.md
# Project Astor

Maven Java project

Astor is a graphical Tango control system administration tool.

* Goal

    * The first goal is to know at a quick glance, if everything is OK in a control system, and otherwise to be able to diagnose a problem and solve it.
    * The second goal is to configure the control system and its components.
    * The third goal is to have long term analysis on components (logs, statistics, usage,?.)

* Principle

    * On each host to be controlled, a device server (called Starter) takes care of all device servers running (or supposed to) on this computer.
    * The controlled server list is read from the TANGO database.
    * A graphical client (called Astor) is connected to all Starter servers and is able to:
    * Display the control system status and component status using coloured icons.
    * Execute actions on components (start, stop, test, configure, display information, ?.
    * Execute diagnostics on components.
    * Execute global analysis on a large number of crates or database.

## Cloning

```
git clone git@github.com:tango-controls/Astor
```

## Documentation 

[ ![Download](https://api.bintray.com/packages/tango-controls/maven/Astor/images/download.svg) ](https://bintray.com/tango-controls/maven/Astor/_latestVersion)
[![Docs](https://img.shields.io/badge/Latest-Docs-orange.svg)](https://tango-controls.readthedocs.io/en/latest/tools-and-extensions/built-in/astor/index.html)


## Building and Installation

### Dependencies

#### Project Dependencies 

* DBBench.jar
* Jive.jar
* JTango.jar
* ATKCore.jar
* ATKWidget.jar
* atkpanel.jar
  

#### Toolchain Dependencies 

* javac 7 or higher
* maven
  

### Build


Instructions on building the project.

```
cd Astor
mvn package
```



CHANGELOG.md
# Astor Changelog

#### Astor-7.3.8 - 15/03/21:
    Remove splash screen replace by a bottom bar

#### Astor-7.3.7 - 04/03/21:
    Update Images and Logo

#### Astor-7.3.6 - 04/03/21:
    Improve Device HierarchyDialog (also used in Jive)

#### Astor-7.3.5 - 24/08/20:
    Fix a bug in case of a server is not defined in database

#### Astor-7.3.4 - 29/06/20:
    Fix a bug in HostInfoDialog refreshing

#### Astor-7.3.2 - 20/05/20:
    Fix a bug in host state calculation

#### Astor-7.3.1 - 21/04/20 (Removed):
    Remove all notifd references

#### Astor-7.3.0 - 20/04/20:
    Does not manage notifd anymore

#### Astor-7.2.10 - 21/11/19:
    Fix a bug in Starter test action

#### Astor-7.2.9 -  25/09/19:
    Fix a bug in TableViewer to test device
    Improve MultiServerCommand dialog

#### Astor-7.2.8 -  28/08/19:
    Improve check states dialog (sorted by class in a tabbed pane)
    
#### Astor-7.2.7 -  19/06/19:
    Improve device selection (sorted by class)

#### Astor-7.2.6 -  12/05/19:
    Remove warning trace

#### Astor-7.2.5 -  02/04/19:
    Minor changes

#### Astor-7.2.4 -  17/01/19:
    Fix a small problem in server startup level

#### Astor-7.2.3 -  09/11/18:
    Fix a bug in host ordering

#### Astor-7.2.2 -  05/11/18:
    Get Starter list from class name.
    Improve README.md.
    Replace ReleaseNote file by CHANGELOG.md

#### Astor-7.2.1 -  09/01/18:
    Add management of a new Starter state (STANDBY) when a server is blocked for a while.

#### Astor-7.1.2 -  04/09/17:
    Add a global check for several instances for a server in servers statistics.

#### Astor-7.1.1 -  31/08/17:
    add a missing icon file

#### Astor-7.1.0 -  31/08/17:
    Manage the Starter check for several instances for a server.

#### Astor-7.0.14 -  04/05/17:
    Update the release number in display.

#### Astor-7.0.13 -  27/04/17:
    Fix a NPE

#### Astor-7.0.12 -  16/02/17:
    Improve check host name when create/clone host object

#### Astor-7.0.11 -  01/02/17:
    Improve startup, level management when start more than 4 servers.

#### Astor-7.0.10 -  16/01/17:
    Fix bug in Start/Stop all servers for several hosts

#### Astor-7.0.9 -  09/01/17:
    Fix a bug in Change TANGO_HOST

#### Astor-7.0.8 -  06/01/17:
    Add DBBench call to monitor database.

#### Astor-7.0.7 -  04/01/17:
    Fix problem in TANGO release
    Check if host is alive before openning a terminal

#### Astor-7.0.6 -  15/12/16:
    Remove old CVS references
    Improve resize host info dialog.
    fix a problem in help for server TANGO release

#### Astor-7.0.5 -  02/12/16:
    Fix a bug in launching JSSHTerminal.

#### Astor-7.0.4 -  01/12/16:
    JSSHTerminal is now used for remote login (JSSHTerminal.jar must be added in CLASSPATH)

#### Astor-7.0.3 -  28/11/16:
    Add a button to start/stop servers on all levels.

#### Astor-7.0.2 -  24/11/16:
    Fix a bug in last branches management.

#### Astor-7.0.1 -  18/11/16:
    Maven project

#### Astor-7.0.0 -  26/08/16:
    Java-7 syntax compatibility.

#### Astor-6.7.1 -  28/06/16:
    Fix a little bug in startup level management.

#### Astor-6.7.0 -  27/01/16:
    Starter domain name could be set with a class property.
    Improve "Tango release for servers" feature

#### Astor-6.6.6 -  30/06/15:
    Java-7 compatibility.
    Manage polling for Tango-9 feature.
    Improve Tango release for servers.
    Add a compatibility to read JTango release.
    Update file headers.

#### Astor-6.6.5 -  08/01/15:
    Improve Server Usage utility

#### Astor-6.6.4 -  08/01/15:
    Improve the statup level change.

#### Astor-6.6.3 -  17/12/14:
    Several invokeLater() methods added.

#### Astor-6.6.2 -  18/11/14:
    ULong management added to event tester.

#### Astor-6.6.1 -  04/11/14:
    Faulty host list added to help menu.

#### Astor-6.6.0 -  01/10/14:
    Server Tango release manage Tango-9 and IDL5

#### Astor-6.5.1 -  05/08/14:
    Improve Last Branches management.

#### Astor-6.5.0 -  19/04/14:
    Manage OFF state for Starter devices.
    Improve DbPollPanel class.

#### Astor-6.4.6 -  29/01/14:
    Package ctrl_system_info added to the jar file.

#### Astor-6.4.5 -  10/01/14:
    Fix a problem on starter properties (HostUsage and Family)
    Clean compilation warnings.

#### Astor-6.4.3 -  07/01/14:
    Remove warning on tools launch.

#### Astor-6.4.2 -  29/10/13:
    Try to fix infinite ping loop.

#### Astor-6.4.1 -  25/10/13:
    Replace dancers by official Tango logo.
    Cleanup part of code.

#### Astor-6.4.0 -  24/09/13:
    Fix a problem in Tango releases help.
    fix a problem on resize with java-7

#### Astor-6.3.9 -  30/08/13:
    Fix problem in pool threads management using new syntax.
    Improve sort methods.
    ServerStatePanel is now an astor.tools class (not from tool_panels package)

#### Astor-6.3.7 -  09/08/13:
    Improve UnAvailableHostsDialog class.

#### Astor-6.3.6 -  12/06/13:
    Server list to be started can now be taken from another host.

#### Astor-6.3.5 -  28/05/13:
    Update splash screen with new logos.

#### Astor-6.3.4 -  16/05/13:
    Minor changes.

#### Astor-6.3.3 -  30/04/13:
    Add a tool to export server from database to another one.

#### Astor-6.3.2 -  03/04/13:
    Add a display for statistics reset by host.

#### Astor-6.3.1 -  19/03/13:
    Fix a problem when starter is killed.

#### Astor-6.3.0 -  06/03/13:
    implement READ_ONLY, DB_READ_ONLY, READ_WRITE modes.

#### Astor-6.2.2 -  21/01/13:
    Fix a bug in adding item on TAC

#### Astor-6.2.1 -  14/01/13:
    Pb on notif icon fixed.

#### Astor-6.2.0 -  07/01/13:
    Group management added on TAC panel.

#### Astor-6.1.2 -  26/10/12:
    Fix a problem when too much devices in SubDevices.

#### Astor-6.1.1 -  26/10/12:
    HostInfoDialog can be run in stand alone.

#### Astor-6.1.0 -  24/10/12:
    Tango release for servers tool added.

#### Astor-6.0.6 -  09/10/12:
    Pb with back slash fixed (?)

#### Astor-6.0.5 -  21/08/12:
    Open Jive on a selected server.


DeviceTree.java
//+======================================================================
// $Source$
//
// Project:   Tango
//
// Description:  java source code for the TempClass class definition .
//
// $Author$
//
// $Revision$
//
// $Log$
// Revision 3.13  2006/09/19 13:04:46  pascal_verdier
// Access control manager added.
//
// Revision 3.12  2005/11/24 12:24:57  pascal_verdier
// DevBrowser utility added.
// MkStarter utility added.
//
// Revision 3.11  2005/11/17 12:30:33  pascal_verdier
// Analysed with IntelliJidea.
//
// Revision 3.10  2005/10/20 13:24:49  pascal_verdier
// Screen position management has been changed.
//
// Revision 3.9  2005/03/11 14:07:53  pascal_verdier
// Pathes have been modified.
//
// Revision 3.8  2005/02/21 13:44:16  pascal_verdier
// Display not controled servers in a JTable.
//
// Revision 3.7  2005/02/16 13:41:02  pascal_verdier
// Add controlled servers info in DeviceTree class.
//
// Revision 3.6  2005/01/18 08:48:20  pascal_verdier
// Tools menu added.
// Not controlled servers list added.
//
// Revision 3.5  2004/09/28 07:01:51  pascal_verdier
// Problem on two events server list fixed.
//
// Revision 3.4  2004/07/08 11:22:58  pascal_verdier
// First revision able to use events.
//
// Revision 3.3  2004/05/04 07:05:27  pascal_verdier
// Bug on notify daemon fixed.
// server reconection transparency added.
//
// Revision 3.2  2004/04/13 12:17:29  pascal_verdier
// DeviceTree class uses the new browsing database commands.
//
// Revision 3.1  2003/11/25 15:56:45  pascal_verdier
// Label on hosts added.
// Notifyd begin to be controled.
//
//
// Copyleft 2003 by European Synchrotron Radiation Facility, Grenoble, France
//               All Rights Reversed
//-======================================================================

package admin.astor;

import fr.esrf.Tango.DevFailed;
import fr.esrf.TangoApi.*;
import fr.esrf.TangoDs.Except;

import javax.swing.*;
import javax.swing.event.TreeExpansionEvent;
import javax.swing.event.TreeExpansionListener;
import javax.swing.tree.*;
import java.awt.*;
import java.awt.event.MouseEvent;
import java.util.StringTokenizer;
import java.util.Vector;

public class DeviceTree extends JTree  implements AstorDefs
{
	private	CtrlSystem			cs;
	private Astor				appli;
	private JTextArea			infoLabel;
	private final String[]		collections = { "Controlled Servers", "Not Controlled servers", "Devices" };

	private final int	CTRL_SERVERS   = 0;
	private final int	N_CTRL_SERVERS = 1;
	private final int	DEVICES        = 2;

	private Monitor	monitor;
	boolean canceled = false;

	 //===============================================================
	//===============================================================
	public DeviceTree(Astor appli, Monitor monitor, JTextArea lbl, String title)
	{
		super();
		this.appli   = appli;
		this.monitor = monitor;
		infoLabel    = lbl;
		cs = new CtrlSystem(appli.tree.hosts.length);

		//	Build panel and its tree
		initComponent(title);
	}
	//===============================================================
	//===============================================================
	private void initComponent(String title)
	{
		//Create the nodes.
		DefaultMutableTreeNode root = new DefaultMutableTreeNode(title);
		createNodes(root);

		//	Create the tree that allows one selection at a time.
		getSelectionModel().setSelectionMode(TreeSelectionModel.SINGLE_TREE_SELECTION);

		//	Create Tree and Tree model
		//------------------------------------
		DefaultTreeModel treeModel = new DefaultTreeModel(root);
		setModel(treeModel);

		//Enable tool tips.
		ToolTipManager.sharedInstance().registerComponent(this);

         // Set the icon for leaf nodes.
		setCellRenderer(new TangoRenderer());

		//	Listen for collapse tree
		addTreeExpansionListener(new TreeExpansionListener () {
			public void treeCollapsed(TreeExpansionEvent e) {
				collapsedPerfomed(e);
			}
			public void treeExpanded(TreeExpansionEvent e) {
				//expandedPerfomed(e);
			}
		});
		//	Add Action listener
		//------------------------------------
		addMouseListener (new java.awt.event.MouseAdapter () {
			public void mouseClicked (java.awt.event.MouseEvent evt) {
				treeMouseClicked (evt);
			}
		});
	}
	//===============================================================
	/**
	 * Create nodes for Servers and devices
	 */
	//===============================================================
	private	DefaultMutableTreeNode[] collnodes;
	private void createNodes(DefaultMutableTreeNode root)
	{
		collnodes = new DefaultMutableTreeNode[collections.length];
		for (int i=0 ; i<collections.length ; i++)
		{
			collnodes[i] = new DefaultMutableTreeNode(collections[i]);
			root.add(collnodes[i]);
		}

		try
		{
			//----------------------------
			//	Builds Servers Tree
			//----------------------------
			DefaultMutableTreeNode[]	s_node = new DefaultMutableTreeNode[2];
			double	ratio;
			String[] servers = ApiUtil.get_db_obj().get_server_name_list();
			for (int i=0 ; i<servers.length ; i++)
			{
				ratio = (double)i/servers.length/2;

				//	Add a server node
				s_node[0] = new DefaultMutableTreeNode(servers[i]);
				s_node[1] = new DefaultMutableTreeNode(servers[i]);
				//	get instances
				String[] instances =
					ApiUtil.get_db_obj().get_instance_name_list(servers[i]);

				//	Then add instance nodes
				for (String instance : instances)
				{
					String servname = servers[i] + "/" + instance;
					if ((canceled=monitor.isCanceled())==true)
						return;
					monitor.setProgressValue(ratio, "Building for Server " + servname);
					//	Check if controlled and add to server node
					DbServInfo info = new DbServer(servname).get_info();
					if (info.controlled)
					{
						//	Update ContolSystem info
						cs.nb_instances++;
						DbServer dbs = new DbServer(servname);
						String[]	classes = dbs.get_class_list();
						cs.nb_classes += classes.length;
						for (String classname : classes)
						{
							//	get device list
							String[]	dn = dbs.get_device_name(classname);
							cs.nb_devices += dn.length;
						}

						s_node[CTRL_SERVERS].add(new DefaultMutableTreeNode(instance));
					}
					else
					{
						cs.not_c_instances++;
						s_node[N_CTRL_SERVERS].add(new DefaultMutableTreeNode(instance));
					}
				}
				//	Add server node if has at least one child
				if (s_node[CTRL_SERVERS].getChildCount()>0)
				{
					cs.nb_servers++;
					collnodes[CTRL_SERVERS].add(s_node[CTRL_SERVERS]);
				}
				if (s_node[N_CTRL_SERVERS].getChildCount()>0)
					collnodes[N_CTRL_SERVERS].add(s_node[N_CTRL_SERVERS]);

				Thread.sleep(10);
			}


			//----------------------------
			//	Builds Devices Tree
			//----------------------------
			String[]	domain;
			String[]	family;
			String[]	member;
			//	Query database for devices list and build tree
			//----------------------------------
			Database	dbase = ApiUtil.get_db_obj();
			//	Query database for Domains list
			domain = dbase.get_device_domain("*");
			for (int i=0 ; i<domain.length ; i++)
			{
				ratio = (double)i/domain.length/2 + 0.5;
				DefaultMutableTreeNode	d_node =
						new DefaultMutableTreeNode(domain[i]);
				collnodes[DEVICES].add(d_node);

				//	Query database for Families list
				String	wildcard = domain[i] + "/*";
				family = dbase.get_device_family(wildcard);
				for (String aFamily : family)
				{
					if ((canceled=monitor.isCanceled())==true)
						return;
					monitor.setProgressValue(ratio,
							"Building for Device family " + domain[i] + "/" + aFamily);

					DefaultMutableTreeNode f_node =
							new DefaultMutableTreeNode(aFamily);
					d_node.add(f_node);

					//	Query database for members list
					wildcard = domain[i] + "/" + aFamily + "/*";
					member = dbase.get_device_member(wildcard);
					for (String aMember : member)
					{
						DefaultMutableTreeNode m_node =
								new DefaultMutableTreeNode(aMember);
						f_node.add(m_node);
					}
				}
				Thread.sleep(10);
			}
		}
		catch(DevFailed e)
		{
			app_util.PopupError.show(appli, e);
		}
		catch(InterruptedException e){}
	}

	//======================================================
	/**
	 *	Manage event on clicked mouse on PogoTree object.
	 */
	//======================================================
	private void treeMouseClicked (java.awt.event.MouseEvent evt)
	{
		//	Check if click is on a node
		if (getRowForLocation(evt.getX(), evt.getY())<1)
			return;

		TreePath	path = getPathForLocation(evt.getX(), evt.getY());
		int			mask = evt.getModifiers();
		//	Check if btn1
		//-------------------------------------
		if ((mask & MouseEvent.BUTTON1_MASK)!=0)
		{
			String	devname;
			switch (path.getPathCount())
			{
			//	If device
			case 5:
				//	Get the device name from path
				devname =	getDeviceName(path);
				if (devname==null)	return;
				break;
			//	if server
			case 4:
				String	servname =	getServerName(path);
				if (servname==null)	return;
				devname = "dserver/" + servname;
				break;
			default:
				return;
			}
			try {
				//	Display device Info
				DeviceInfo	info =
					ApiUtil.get_db_obj().get_device_info(devname);
				infoLabel.setText(info.toString());
			}
			catch (DevFailed e) {
				infoLabel.setText(" ");
			}

			//	Popup Host Info Dialog only if double click
			//--------------------------------------------------
			if(evt.getClickCount() == 2)
			{
				showHostInfoDialogForDevice(devname, evt);
			}
		}
	}
	//======================================================
	//======================================================
	public void collapsedPerfomed(TreeExpansionEvent e)
	{
		//	Get path
		TreePath	path = e.getPath();
		if (path.getPathCount()>2)
			return;
		//	Get concerned node
		DefaultMutableTreeNode	node =
			(DefaultMutableTreeNode)path.getPathComponent(path.getPathCount()-1);
		//	do not collapse if Root
		if (path.getPathCount()==1)
		{
			setExpandedState(new TreePath(node.getPath()), true);
			app_util.PopupMessage.showImage(appli, cs.toString(), img_path + "tango_icon.jpg");
		}
	}
	//======================================================
	//======================================================
	String[][] getNotCtrlServers()
	{
		DefaultMutableTreeNode	node = collnodes[N_CTRL_SERVERS];
		int	nb_serv = node.getChildCount();
		Vector	v_serv = new Vector();
		//	Search all Classes
		for (int i=0 ; i<nb_serv ; i++)
		{
			node = node.getNextNode();
			String	servname = node.toString();
			int	nb_inst = node.getChildCount();
			try
			{
				//	For all instances
				for (int j=0 ; j<nb_inst ; j++)
				{
					node = node.getNextNode();
					String		server = servname + "/" + node;
					DbServInfo	info = new DbServer(server).get_info();
					//	Store in vector if controlled
					if (!info.controlled)
						//	Starter cannot controle itself.
						if (!servname.equals("Starter"))
							v_serv.add(server);
				}
			}
			catch(DevFailed e){
				v_serv.add(e.errors[0].desc);
			}
		}

		//	Add last exported date
		String[][]	result = new String[v_serv.size()][];
		for (int i=0 ; i<v_serv.size() ; i++)
		{
			String	server = (String)v_serv.elementAt(i); 
			result[i]    = new String[2];
			result[i][0] = server;
			//	Get last started date
			try
			{
				DeviceInfo	info2 = new DbDevice("dserver/"+server).get_info();
				result[i][1] = info2.last_exported;
			}
			catch (DevFailed e){
				result[i][1] = e.errors[0].desc;
			}
		}

		return result;
	}
	//======================================================
	//======================================================
	@SuppressWarnings({"UNUSED_SYMBOL"})
	private void showHostInfoDialogForDevice(String devname, MouseEvent evt)
	{
		try
		{
			//	Get host name from device
			String	hostname = new IORdump(devname).get_host();
			if (hostname==null)
				Except.throw_exception("UNKNOWN_HOST",
					"May be this device has never been exported !", "");
			//	Take off IP address if exists
			StringTokenizer st = new StringTokenizer(hostname);
			hostname = st.nextToken();
			//	Take off name extention (e.g. .esrf.fr) if exists
			st = new StringTokenizer(hostname, ".");
			hostname = st.nextToken();

			//	Select Host on main Tree and Popup host panel
			appli.tree.setSelectionPath(hostname);
			appli.tree.displayHostInfo();
		}
		catch (DevFailed e)
		{
			app_util.PopupError.show(this, e);
		}
	}
	//======================================================
	//======================================================
	private String getDeviceName(TreePath path)
	{
		//	Check if Device
		String	collec = path.getPathComponent(1).toString();
		if (!collec.equals(collections[DEVICES]))
			return null;

		return
			path.getPathComponent(2).toString() + "/" +
			path.getPathComponent(3).toString() + "/" +
			path.getPathComponent(4).toString();
	}
	//======================================================
	//======================================================
	private String getServerName(TreePath path)
	{
		//	Check if Server
		String	collec = path.getPathComponent(1).toString();
		//if (collec.equals(collections[CTRL_SERVERS])==false)
		if (collec.equals(collections[CTRL_SERVERS]) && collec.equals(collections[N_CTRL_SERVERS]))
			return null;

		return
			path.getPathComponent(2).toString() + "/" +
			path.getPathComponent(3).toString();
	}
	//======================================================
	//======================================================
	String csInfo()
	{
		return cs.toString();
	}
//===============================================================
/**
 *	Control System info class
 */
//===============================================================
	class CtrlSystem
	{
		int	nb_hosts     = 0;
		int	nb_servers   = 0;
		int	nb_instances = 0;
		int	nb_classes   = 0;
		int	nb_devices   = 0;
		int	not_c_instances= 0;

		public CtrlSystem(int nb_hosts)
		{
			this.nb_hosts = nb_hosts;
		}
		public String toString()
		{
			String  str = "";
			str +=  nb_hosts     + "  Hosts controlled.\n";
			str +=  nb_servers   + "  Different controlled servers.\n";
			str +=  nb_instances + "  Controlled servers/instances.\n";
			str +=  nb_classes   + "  Controlled classes.\n";
			str +=  nb_devices   + "  Controlled devices.\n\n";
			str +=  not_c_instances+ "  NOT controlled servers/instances.";
			return str;
		}
	}
	//===============================================================

//===============================================================
/**
 *	Renderer Class
 */
//===============================================================
    private class TangoRenderer extends DefaultTreeCellRenderer
	{
		private ImageIcon	tangoIcon;
		private ImageIcon	serv_icon;
		private ImageIcon	dev_icon;
		private Font[]		fonts;

		private final int	TITLE = 0;
		private final int	LEAF  = 1;

		//===============================================================
		//===============================================================
		public TangoRenderer()
		{
			//tangoIcon = new ImageIcon(getClass().getResource(img_path + "tango_icon.jpg"));
			tangoIcon = new ImageIcon(getClass().getResource(img_path + "network5.gif"));

			serv_icon = new ImageIcon(getClass().getResource(img_path + "server.gif"));
			dev_icon = new ImageIcon(getClass().getResource(img_path + "device.gif"));
			fonts = new Font[2];
			fonts[TITLE] = new Font("helvetica", Font.BOLD, 18);
			fonts[LEAF]  = new Font("helvetica", Font.PLAIN, 12);
		}

		//===============================================================
		//===============================================================
		public Component getTreeCellRendererComponent(
                            JTree tree,
                            Object obj,
                            boolean sel,
                            boolean expanded,
                            boolean leaf,
                            int row,
                            boolean hasFocus) {

            super.getTreeCellRendererComponent(
                            tree, obj, sel,
                            expanded, leaf, row,
                            hasFocus);

 			setBackgroundNonSelectionColor(java.awt.Color.white);
			if (row==0)
			{
				//	ROOT
 				setBackgroundSelectionColor(java.awt.Color.white);
				setIcon(tangoIcon);
				setFont(fonts[TITLE]);
				setToolTipText("Double click to popup info");
			}
			else
			{
				DefaultMutableTreeNode node = (DefaultMutableTreeNode)obj;
				Object user_obj = node.getUserObject();
				//setIcon(node_icon);
				switch (node.getLevel())
				{
				case 1:
 					setBackgroundSelectionColor(java.awt.Color.white);
					setToolTipText("Tango " + user_obj.toString());
					setFont(fonts[TITLE]);
					break;
				default:
 					setBackgroundSelectionColor(java.awt.Color.lightGray);
					if (leaf)
					{
						//	Check if collection is device or server
						//	to display associated icon
						TreeNode[]	path = node.getPath();
						DefaultMutableTreeNode c_node =
									(DefaultMutableTreeNode)path[1];
						String collec = (c_node.getUserObject()).toString();
						if (collec.equals(collections[DEVICES]))
							setIcon(dev_icon);
						else
							setIcon(serv_icon);
						setToolTipText("Double click to popup host window.");
					}
					setFont(fonts[LEAF]);
					break;
				}
			}
            return this;
        }
	}//	End of Renderer Class
}

